### ISSUE TITLE
C++: MSVC Driver: Properly predefine language version macros

### ISSUE THREAD

**Issue by duncanp-sonar** [SonarSource/sonarlint-visualstudio#1106]
### Description

SLVS contains a port of the sonar-cpp Java `MSVCDriver` class.
Bug [CPP-2305](https://jira.sonarsource.com/browse/CPP-2305) was found and fixed on the Java side. [The fix](https://github.com/SonarSource/sonar-cpp/pull/670) needs to be ported to SLVS.

### Repro steps

1. Configure the project to use C++ Language Standard `ISO C++17 Standard (std:c++17)`

2. Analyze any code that uses C++17 standard library. Example:
```
#include<optional>
// std::optional is introduced in c++17.
std::optional<int> foo(int x) // Parsing error issue raised.
{
  if (x < 0)
    return std::nullopt; // Parsing error issue raised.
  return x;
};
```
This code will trigger parsing error issue and will impact the analysis results.

NOTE: to see the parsing errors reported as issues you need to enabled the hidden `cpp:ParsingError` rule in the `settings.json` file:

```
{
  "sonarlint.rules": {
    "cpp:ParsingError": {
      "Level": "On"
    }
  }
}
```

### Expected behavior

File is analysed successfully and all issues reported.

### Actual behavior

Not all issues are reported.

Any parsing errors are swallowed by the `subprocess.exe`. No errors or warnings are reported in the UI or the SonarLint output pane i.e. there is no indication that there were any problems during analysis or that there may be missing issues.

Other issues in the same file in code that can be parsed will still be reported, whether they occur before or after the code that cannot be parsed.

### Known workarounds

None

### Related information

* SonarLint for VS Version: v4.x+
* Visual Studio Version: v2019 - v2017
* Languages in your solution: C++17

**Comment by duncanp-sonar** [comment]
Re-opening: the modified code fails on VS2015 with the following error message:
```
Unable to collect C/C++ configuration for c:\Data\Code\CppWin32Project1\CppWin32Project1\CppWin32Project1.cpp: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> Microsoft.VisualStudio.ProjectSystem.ProjectException: The "CL" rule is missing the "LanguageStandard" property.
   at Microsoft.VisualStudio.ProjectSystem.Utilities.ProjectErrorUtilities.ThrowProjectExceptionHelper(Exception innerException, String unformattedMessage, Object[] args)
   at Microsoft.VisualStudio.ProjectSystem.Utilities.ProjectErrorUtilities.ThrowProjectException(String message, Object arg0, Object arg1)
   at Microsoft.VisualStudio.Project.VisualC.VCProjectEngine.VCRulePropertyStorage.<>c__DisplayClass5_0.<<GetEvaluatedPropertyValue>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Project.VisualC.VCProjectEngine.ApartmentMarshaler.<>c__DisplayClass9_0`1.<<Invoke>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Threading.JoinableTask.CompleteOnCurrentThread()
   at Microsoft.VisualStudio.Threading.JoinableTask`1.CompleteOnCurrentThread()
   at Microsoft.VisualStudio.Threading.JoinableTaskFactory.Run[T](Func`1 asyncMethod, JoinableTaskCreationOptions creationOptions)
   at Microsoft.VisualStudio.Threading.JoinableTaskFactory.Run[T](Func`1 asyncMethod)
   at Microsoft.VisualStudio.Project.VisualC.VCProjectEngine.ApartmentMarshaler.Invoke[TResult](Func`1 method)
   at Microsoft.VisualStudio.Project.VisualC.VCProjectEngine.VCRulePropertyStorage.GetEvaluatedPropertyValue(String propertyName)
   at Microsoft.VisualStudio.Project.VisualC.VCProjectEngine.VCToolBase.Microsoft.VisualStudio.VCProjectEngine.IVCRulePropertyStorage.GetEvaluatedPropertyValue(String propertyName)
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at SonarLint.VisualStudio.Integration.Vsix.CFamily.CFamilyHelper.FileConfig.GetEvaluatedPropertyValue(Object fileTool, String propertyName)
   at System.Dynamic.UpdateDelegates.UpdateAndExecute3[T0,T1,T2,TRet](CallSite site, T0 arg0, T1 arg1, T2 arg2)
   at SonarLint.VisualStudio.Integration.Vsix.CFamily.CFamilyHelper.FileConfig.TryGet(ProjectItem projectItem, String absoluteFilePath)
   at SonarLint.VisualStudio.Integration.Vsix.CFamily.CFamilyHelper.TryGetConfig(ILogger logger, ProjectItem projectItem, String absoluteFilePath)
```
